lesson20.md

```diff
@@ -0,0 +1,272 @@
+map.setView([52.531587, 13.384742], 15);
+```
+
+---
+
+### Leaflet
+
+* Leaflet itself doesn't come with data
+* We can use Mapbox data (which comes from OpenStreetMap)
+* Note - Mapbox is not free, but requires an access token. Check the slack  channel and set it in the `accessToken` property below:
+
+```js
+let layer = L.tileLayer('https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token={accessToken}', {
+    attribution: 'Map data &copy; <a href="https://www.openstreetmap.org/">OpenStreetMap</a> contributors, <a href="https://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery © <a href="https://www.mapbox.com/">Mapbox</a>',
+    maxZoom: 18,
+    id: 'mapbox/streets-v11',
+    tileSize: 512,
+    zoomOffset: -1,
+    accessToken: 'your.mapbox.access.token' // YOUR TOKEN HERE!
+});
+layer.addTo(map);
+```
+
+---
+
+### Add a marker
+
+* We can add markers to the map:
+
+```js
+let marker = L.marker([52.531587, 13.384742]);
+marker.addTo(map);
+```
+
+---
+
+### Add a popup to the marker
+
+```js
+let popup = marker.bindPopup("<b>This is ReDI!</b>");
+popup.openPopup();
+```
+
+---
+
+### Add a listener
+
+```js
+function onMapClick(e) {
+    console.log("You clicked the map at " + e.latlng);
+}
+map.on('click', onMapClick);
+```
+
+---
+
+### Let's play
+
+* Add a couple of more markers, e.g. your home, your favorite places in Berlin
+* Can you add a button below the `div` that sets the view of the map to your favorite place? (hint: use `map.setView()` function in your button's `onclick` function)
+
+---
+
+### Homework
+
+* Create a `<select id="countries" onchange="countryChanged()"></select>` element in your HTML page
+* Call the restcountries.eu "all" API: https://restcountries.eu/rest/v2/all
+* For every country, create one `<option>` element using `document.createElement()`
+* In your `countryChanged()` function, set the view of your Leaflet map to that country's latitude and longitude
+* Try it on your own. If you get stuck, check the hints on the next slides
+* Still stuck? Ask on slack, please :)
+
+---
+
+### Homework Hints
+
+First, we need to fetch all countries. Remember the `fetch` API?
+
+```js
+async function fetchCountries() {
+    let response = await fetch("https://restcountries.eu/rest/v2/all");
+    let countries = await response.json();
+
+    // now populate the <select> element! (see next slide for hint)
+}
+```
+
+---
+
+### Homework Hints
+
+Let's populate our `<select>` element:
+
+```js
+// inside your fetchCountries() function:
+let countriesElement = document.getElementById("countries");
+
+for (let country of countries) {
+    let countryElement = document.createElement("option");
+    // put the name of the country as textContent
+    countryElement.textContent = country.name;
+    // put the latitude and longitude as the option's value attribute
+    // since value must be a string, let's convert the latlng array to JSON:
+    countryElement.value = JSON.stringify(country.latlng);
```

Comment:
Explain why we are stringifying the latlng property. Maybe even show the typeof country.latlng before doing the conversion.